<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 Quality Control | Models</title>
  <meta name="description" content="This report covers all classes from PLPTH 820 course." />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="3 Quality Control | Models" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This report covers all classes from PLPTH 820 course." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Quality Control | Models" />
  
  <meta name="twitter:description" content="This report covers all classes from PLPTH 820 course." />
  

<meta name="author" content="Gabriela Romero Campos" />


<meta name="date" content="2025-05-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-acquisition.html"/>
<link rel="next" href="pseudosamples-from-raw-reads.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.10.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-2.5.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">AGRO 920</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#data"><i class="fa fa-check"></i><b>1.1</b> Data</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#project-overview"><i class="fa fa-check"></i><b>1.2</b> Project Overview</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#objective"><i class="fa fa-check"></i><b>1.2.1</b> Objective</a></li>
<li class="chapter" data-level="1.2.2" data-path="index.html"><a href="index.html#why-k-mers"><i class="fa fa-check"></i><b>1.2.2</b> Why k-mers?</a></li>
<li class="chapter" data-level="1.2.3" data-path="index.html"><a href="index.html#why-pseudo-samples-ps"><i class="fa fa-check"></i><b>1.2.3</b> Why pseudo-samples (PS)?</a></li>
<li class="chapter" data-level="1.2.4" data-path="index.html"><a href="index.html#why-machine-learning"><i class="fa fa-check"></i><b>1.2.4</b> Why machine learning?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-acquisition.html"><a href="data-acquisition.html"><i class="fa fa-check"></i><b>2</b> Data Acquisition</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-acquisition.html"><a href="data-acquisition.html#download"><i class="fa fa-check"></i><b>2.1</b> Download</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#fastqc-and-multiqc"><i class="fa fa-check"></i><b>3.1</b> FastQC and MultiQC</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#duplicate-removal-with-markduplicates"><i class="fa fa-check"></i><b>3.2</b> Duplicate Removal with MarkDuplicates</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="pseudosamples-from-raw-reads.html"><a href="pseudosamples-from-raw-reads.html"><i class="fa fa-check"></i><b>4</b> Pseudosamples from Raw Reads</a>
<ul>
<li class="chapter" data-level="4.1" data-path="pseudosamples-from-raw-reads.html"><a href="pseudosamples-from-raw-reads.html#merge-paired-end-reads-into-a-single-fasta"><i class="fa fa-check"></i><b>4.1</b> Merge Paired-end Reads into a Single FASTA</a></li>
<li class="chapter" data-level="4.2" data-path="pseudosamples-from-raw-reads.html"><a href="pseudosamples-from-raw-reads.html#split-into-pseudosamples"><i class="fa fa-check"></i><b>4.2</b> Split into Pseudosamples</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html"><i class="fa fa-check"></i><b>5</b> K-mer Extraction and Filtering</a>
<ul>
<li class="chapter" data-level="5.1" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#k-mer-counting-with-jellyfish"><i class="fa fa-check"></i><b>5.1</b> K-mer Counting with Jellyfish</a></li>
<li class="chapter" data-level="5.2" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#pseudosample-merging-by-treatment"><i class="fa fa-check"></i><b>5.2</b> Pseudosample Merging by Treatment</a></li>
<li class="chapter" data-level="5.3" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#global-merging-by-group"><i class="fa fa-check"></i><b>5.3</b> Global Merging by Group</a></li>
<li class="chapter" data-level="5.4" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#aggregation-and-total-count-per-k-mer"><i class="fa fa-check"></i><b>5.4</b> Aggregation and Total Count per K-mer</a></li>
<li class="chapter" data-level="5.5" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#filtering-kmers"><i class="fa fa-check"></i><b>5.5</b> Filtering kmers</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#step-1-abundance-threshold"><i class="fa fa-check"></i><b>5.5.1</b> Step 1: Abundance Threshold</a></li>
<li class="chapter" data-level="5.5.2" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#step-2-coefficient-of-variation-cv"><i class="fa fa-check"></i><b>5.5.2</b> Step 2: Coefficient of Variation (CV)</a></li>
<li class="chapter" data-level="5.5.3" data-path="k-mer-extraction-and-filtering.html"><a href="k-mer-extraction-and-filtering.html#step-3-redundancy-removal-with-cd-hit-est"><i class="fa fa-check"></i><b>5.5.3</b> Step 3: Redundancy Removal with CD-HIT-EST</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="exploratory-analysis.html"><a href="exploratory-analysis.html"><i class="fa fa-check"></i><b>6</b> Exploratory Analysis</a>
<ul>
<li class="chapter" data-level="6.1" data-path="exploratory-analysis.html"><a href="exploratory-analysis.html#total-counts"><i class="fa fa-check"></i><b>6.1</b> Total counts</a></li>
<li class="chapter" data-level="6.2" data-path="exploratory-analysis.html"><a href="exploratory-analysis.html#variance"><i class="fa fa-check"></i><b>6.2</b> Variance</a></li>
<li class="chapter" data-level="6.3" data-path="exploratory-analysis.html"><a href="exploratory-analysis.html#pca"><i class="fa fa-check"></i><b>6.3</b> PCA</a></li>
<li class="chapter" data-level="6.4" data-path="exploratory-analysis.html"><a href="exploratory-analysis.html#t-sne"><i class="fa fa-check"></i><b>6.4</b> t-SNE</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="models.html"><a href="models.html"><i class="fa fa-check"></i><b>7</b> Models</a>
<ul>
<li class="chapter" data-level="7.1" data-path="models.html"><a href="models.html#random-forest"><i class="fa fa-check"></i><b>7.1</b> Random Forest</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="models.html"><a href="models.html#confusion-matrices"><i class="fa fa-check"></i><b>7.1.1</b> Confusion Matrices</a></li>
<li class="chapter" data-level="7.1.2" data-path="models.html"><a href="models.html#performance-metrics-and-roc-curves"><i class="fa fa-check"></i><b>7.1.2</b> Performance Metrics and ROC Curves</a></li>
<li class="chapter" data-level="7.1.3" data-path="models.html"><a href="models.html#important-k-mers-associated-with-heat-stress-1h"><i class="fa fa-check"></i><b>7.1.3</b> Important k-mers associated with heat stress (1h)</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Models</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="quality-control" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">3</span> Quality Control<a href="quality-control.html#quality-control" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Although this RNA-seq dataset came from a published study, it didn’t include any quality control metrics. So, I ran an independent QC check to make sure the reads was well. This included <code>FastQC</code>, <code>MultiQC</code>, and duplicate removal with <code>MarkDuplicates</code> from the Picard toolkit (all with the goal of reducing technical noise and <em>preserving my sanity</em> for the analyses ahead).</p>
<div id="fastqc-and-multiqc" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> FastQC and MultiQC<a href="quality-control.html#fastqc-and-multiqc" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Each paired-end FASTQ file was evaluated using FastQC (v0.12.1) via SLURM array jobs. After that, I used MultiQC to bring everything together into one readable summary.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="quality-control.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash -l</span></span>
<span id="cb2-2"><a href="quality-control.html#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --job-name=data920_fastqc</span></span>
<span id="cb2-3"><a href="quality-control.html#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=02:00:00</span></span>
<span id="cb2-4"><a href="quality-control.html#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=50G</span></span>
<span id="cb2-5"><a href="quality-control.html#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb2-6"><a href="quality-control.html#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --ntasks-per-node=10</span></span>
<span id="cb2-7"><a href="quality-control.html#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --array=1-14</span></span>
<span id="cb2-8"><a href="quality-control.html#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="quality-control.html#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Load Fastqc </span></span>
<span id="cb2-10"><a href="quality-control.html#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load FastQC/0.12.1-Java-11</span>
<span id="cb2-11"><a href="quality-control.html#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="quality-control.html#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># finding R1 and R2 files</span></span>
<span id="cb2-13"><a href="quality-control.html#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="va">FASTQ_FILES_R1</span><span class="op">=</span><span class="va">($(</span><span class="fu">ls</span> /homes/grcampos/AGRO920/data/<span class="pp">*</span>_1.fastq.gz<span class="va">))</span></span>
<span id="cb2-14"><a href="quality-control.html#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="va">FASTQ_FILES_R2</span><span class="op">=</span><span class="va">($(</span><span class="fu">ls</span> /homes/grcampos/AGRO920/data/<span class="pp">*</span>_2.fastq.gz<span class="va">))</span></span>
<span id="cb2-15"><a href="quality-control.html#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="quality-control.html#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># One file per array index</span></span>
<span id="cb2-17"><a href="quality-control.html#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="va">FILE_R1</span><span class="op">=</span><span class="va">${FASTQ_FILES_R1</span><span class="op">[</span><span class="va">$SLURM_ARRAY_TASK_ID</span> - 1<span class="op">]</span><span class="va">}</span> <span class="co">#bash arrays starts with 0</span></span>
<span id="cb2-18"><a href="quality-control.html#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="va">FILE_R2</span><span class="op">=</span><span class="va">${FASTQ_FILES_R2</span><span class="op">[</span><span class="va">$SLURM_ARRAY_TASK_ID</span> - 1<span class="op">]</span><span class="va">}</span></span>
<span id="cb2-19"><a href="quality-control.html#cb2-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-20"><a href="quality-control.html#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Running Fatsqc</span></span>
<span id="cb2-21"><a href="quality-control.html#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing </span><span class="va">$FILE_R1</span><span class="st"> and </span><span class="va">$FILE_R2</span><span class="st">&quot;</span></span>
<span id="cb2-22"><a href="quality-control.html#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-t</span> 10 <span class="at">-o</span> /homes/grcampos/AGRO920/fastqc_reports <span class="st">&quot;</span><span class="va">$FILE_R1</span><span class="st">&quot;</span> <span class="st">&quot;</span><span class="va">$FILE_R2</span><span class="st">&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="quality-control.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load MultiQC/1.14-foss-2022a</span>
<span id="cb3-2"><a href="quality-control.html#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="quality-control.html#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> <span class="at">--job-name</span><span class="op">=</span>mqc920 <span class="at">--time</span><span class="op">=</span>02:00:00 <span class="at">--mem</span><span class="op">=</span>10G <span class="at">--nodes</span><span class="op">=</span>1 <span class="at">--wrap</span><span class="op">=</span><span class="st">&quot;multiqc /bulk/eakhunov/grcampos/embeddings/fastqc_reports -o /bulk/eakhunov/grcampos/embeddings/fastqc_reports&quot;</span></span></code></pre></div>
<p><strong>Output:</strong> All FastQC and MultiQC reports are saved in the <code>fastqc_reports/</code> directory.</p>
<p><a href="files/multiqc_report.html" target="_blank">Click here to open the MultiQC report</a></p>
</div>
<div id="duplicate-removal-with-markduplicates" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Duplicate Removal with MarkDuplicates<a href="quality-control.html#duplicate-removal-with-markduplicates" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Next, I addressed the possibility of overrepresented sequences, which could skew k-mer frequencies. Using Picard’s MarkDuplicates, I converted the FASTQ files to BAM, marked duplicates, converted them back to deduplicated FASTQ files, and cleaned up all the temporary BAMs to avoid filling up the server (again).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="quality-control.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash -l</span></span>
<span id="cb4-2"><a href="quality-control.html#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --job-name=data920_markD</span></span>
<span id="cb4-3"><a href="quality-control.html#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=4-00:00:00</span></span>
<span id="cb4-4"><a href="quality-control.html#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=50G</span></span>
<span id="cb4-5"><a href="quality-control.html#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb4-6"><a href="quality-control.html#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --ntasks-per-node=10</span></span>
<span id="cb4-7"><a href="quality-control.html#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --array=1-14</span></span>
<span id="cb4-8"><a href="quality-control.html#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="quality-control.html#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load picard/2.25.1-Java-11</span>
<span id="cb4-10"><a href="quality-control.html#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="quality-control.html#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co">#  List of paired-end FASTQ files</span></span>
<span id="cb4-12"><a href="quality-control.html#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="va">FASTQ_FILES_R1</span><span class="op">=</span><span class="va">($(</span><span class="fu">ls</span> data/<span class="pp">*</span>_1.fastq.gz<span class="va">))</span></span>
<span id="cb4-13"><a href="quality-control.html#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="va">FASTQ_FILES_R2</span><span class="op">=</span><span class="va">($(</span><span class="fu">ls</span> data/<span class="pp">*</span>_2.fastq.gz<span class="va">))</span></span>
<span id="cb4-14"><a href="quality-control.html#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="quality-control.html#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co"># One file per array index</span></span>
<span id="cb4-16"><a href="quality-control.html#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="va">FILE_R1</span><span class="op">=</span><span class="va">${FASTQ_FILES_R1</span><span class="op">[</span><span class="va">$SLURM_ARRAY_TASK_ID</span> - 1<span class="op">]</span><span class="va">}</span></span>
<span id="cb4-17"><a href="quality-control.html#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="va">FILE_R2</span><span class="op">=</span><span class="va">${FASTQ_FILES_R2</span><span class="op">[</span><span class="va">$SLURM_ARRAY_TASK_ID</span> - 1<span class="op">]</span><span class="va">}</span></span>
<span id="cb4-18"><a href="quality-control.html#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLE_NAME</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="st">&quot;</span><span class="va">$FILE_R1</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/_1.fastq.gz//&#39;</span><span class="va">)</span></span>
<span id="cb4-19"><a href="quality-control.html#cb4-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-20"><a href="quality-control.html#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing sample: </span><span class="va">$SAMPLE_NAME</span><span class="st">&quot;</span></span>
<span id="cb4-21"><a href="quality-control.html#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="ex">java</span> <span class="at">-jar</span> <span class="va">$EBROOTPICARD</span>/picard.jar FastqToSam <span class="at">-F1</span><span class="op">=</span><span class="st">&quot;</span><span class="va">$FILE_R1</span><span class="st">&quot;</span> <span class="at">-F2</span><span class="op">=</span><span class="st">&quot;</span><span class="va">$FILE_R2</span><span class="st">&quot;</span> O=<span class="st">&quot;bam_files/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_unaligned.bam&quot;</span> <span class="at">-SM</span><span class="op">=</span><span class="st">&quot;</span><span class="va">$SAMPLE_NAME</span><span class="st">&quot;</span></span>
<span id="cb4-22"><a href="quality-control.html#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;FASTQ to BAM completed for </span><span class="va">$SAMPLE_NAME</span><span class="st">&quot;</span></span>
<span id="cb4-23"><a href="quality-control.html#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="quality-control.html#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove duplicates with MarkDuplicates</span></span>
<span id="cb4-25"><a href="quality-control.html#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="ex">java</span> <span class="at">-jar</span> <span class="va">$EBROOTPICARD</span>/picard.jar MarkDuplicates I=<span class="st">&quot;bam_files/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_unaligned.bam&quot;</span> O=<span class="st">&quot;bam_files/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_deduplicated.bam&quot;</span> M=<span class="st">&quot;bam_files/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_duplicate_metrics.txt&quot;</span> REMOVE_DUPLICATES=true VALIDATION_STRINGENCY=LENIENT <span class="co">#Lenient: Display warnings but will continue processing even with minor errors.</span></span>
<span id="cb4-26"><a href="quality-control.html#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;MarkDuplicates completed for </span><span class="va">$SAMPLE_NAME</span><span class="st">&quot;</span></span>
<span id="cb4-27"><a href="quality-control.html#cb4-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-28"><a href="quality-control.html#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert BAM back to FASTQ</span></span>
<span id="cb4-29"><a href="quality-control.html#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="ex">java</span> <span class="at">-jar</span> <span class="va">$EBROOTPICARD</span>/picard.jar SamToFastq I=<span class="st">&quot;bam_files/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_deduplicated.bam&quot;</span> F=<span class="st">&quot;fastq_filtered/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_R1_clean.fastq.gz&quot;</span> F2=<span class="st">&quot;/fastq_filtered/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_R2_clean.fastq.gz&quot;</span></span>
<span id="cb4-30"><a href="quality-control.html#cb4-30" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;BAM to FASTQ completed for </span><span class="va">$SAMPLE_NAME</span><span class="st">&quot;</span></span>
<span id="cb4-31"><a href="quality-control.html#cb4-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-32"><a href="quality-control.html#cb4-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove BAM files</span></span>
<span id="cb4-33"><a href="quality-control.html#cb4-33" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="st">&quot;bam_files/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_unaligned.bam&quot;</span></span>
<span id="cb4-34"><a href="quality-control.html#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="st">&quot;bam_files/</span><span class="va">${SAMPLE_NAME}</span><span class="st">_deduplicated.bam&quot;</span></span>
<span id="cb4-35"><a href="quality-control.html#cb4-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-36"><a href="quality-control.html#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Processing completed for </span><span class="va">$SAMPLE_NAME</span><span class="st">!&quot;</span></span></code></pre></div>
<p><strong>Output:</strong> All and duplication metrics (<code>*_duplicate_metrics.txt</code>) were saved to the <code>bam_files/</code> directory. The intermediate BAM files were removed to save memory.</p>
<p>No major PCR duplication was observed. According to Picard, most samples had zero or negligible duplicated reads, meaning the original RNA-seq libraries were solid.</p>
<blockquote>
<p>Note: All scripts used in this section are stored in the <code>scripts/</code> directory as <code>02_runFastqc.sbatch</code>, and <code>03_runDeduplication.sbatch</code>.</p>
</blockquote>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-acquisition.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="pseudosamples-from-raw-reads.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/02_Quality_Control.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
},
"self_contained": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
